### ISSUE TITLE
Leaving a limit switch

### ISSUE THREAD

**Issue by srgblnch** [sardana-org/sardana#507]
Before to produce any proposal I like to discuss it first. I also like to know if someone else have had this behaviour to evolve in the best solution for all of us. I did a _fast_ check in the open issues and I didn't see something like what I wanna mention already open.

I have a [motor](https://sourceforge.net/p/sardana/controllers.git/ci/master/tree/python/motor/PmacCtrl/AlbaLtpTurboPmacCtrl.py) with magnetic limit switches that needs a bit of time to release the limit when leaving in the opposite direction movement (it's about a millimetre).

From one part of the code, the [PoolMotor](https://github.com/sardana-org/sardana/blob/develop/src/sardana/pool/poolmotor.py) will be in ALARM state when any limit switch is "press". When one moves the motor in the opposite direction, the [PoolMotion](https://github.com/sardana-org/sardana/blob/develop/src/sardana/pool/poolmotion.py) will check in the _action_loop()_ the state of the motor object and abort the motion if it says alarm.

Did you see already the issue? If the limit has not been released by when the motionObj checks it, the movement will be aborted even it is the only way to leave the limit switch.

* Did someone else solve it somehow?
* If not, how would be better?
    1. Modify the _PoolMotor_ to know motion direction when _position_ is set, to avoid reporting alarm when releasing the limit?
    2. Modify the _PoolMotion_ to, in case of alarm is reported, check the cause of the alarm to decide if it is a reason to abort the movement?
    3. Any other proposal?

**Comment by teresanunez** [comment]
Hi, I assume you do not have a Tango Device behind the Sardana Controller. In DESY our Motor Controllers always talk to Tango Devices and those are the ones talking to the hardware. We allow the Tango Device to move the motor in case of Limit Switch, so we can go out from the ALARM state, and then go back to Sardana. But this is not a 'Sardana solution'. Looking at your proposed solutions, I think the only difference that they imply is to keep the ALARM state during the release from the switch or not:
solution ii should also know about the direction of the motion to decide if the movement has to be aborted or not.

**Comment by reszelaz** [comment]
First of all, we should clarify if the Sardana motor should return the ALARM state in case a limit switch is active. After checking some of the controllers in the 3rd-party repository I have found that only the IcePAP and Pmac controller do that. But it is better if their developers reply here cause I may be wrong. Furthermore, the [present documentation](http://www.sardana-controls.org/en/latest/devel/howto_controllers/howto_motorcontroller.html) does not say anything about it. But I found it explicitly said in the [old documentation](http://www.sardana-controls.org/en/latest/devel/api/tango_device_pool.html?highlight=limit#the-states).

Apart from the options proposed by @srgblnch I see one more. To disable the emergency stop feature per motor/motor group/pseudo motor either for a single move or temporarily e.g. 5, 10, ... seconds. This could be a boolean attribute associated with this object.

**Comment by teresanunez** [comment]
Hi @reszelaz, you are right, the motors at DESY do not return ALARM State in case of limit switch,
 they return the right switch state, but the motor stays in ON. We rely on our tango devices behind for
 solving these kind of problems.  But well, we could check what happens if we change this.
 With the option you propose, do you mean to allow a movement for 5-10 seconds even if the motor is
in alarm state and interrupt it after this time if the state remains in ALARM?.

**Comment by reszelaz** [comment]
> do you mean to allow a movement for 5-10 seconds even if the motor is in alarm state and interrupt it after this time if the state remains in ALARM?

Yes, this is a similar approach that we apply in the case of the hardware anti-collision switches that are controlled by a PLC and the IcePAP disable signal. The PLC ignores the anti-collision signal during a while in order to allow the IcePAP to move.

At the moment I like the most the following option. Modify the IcePAP and Pmac controllers so they return the ALARM state in case a limit switch is active. Then, enhance the emergency stop feature with the analysis of the limit switch state and the motion direction so it aborts the motion in case it is moving in the positive direction and the positive limit is active or in the negative direction and the negative limit is active. Of course, I would still abort the mtion if the state is ALARM. But now the state would not be linked with the limit switches.

**Comment by teresanunez** [comment]
@reszelaz, I agree with your solution. But you have make a mistake in the sentence:
"Modify the IcePAP and Pmac controllers so they return the ALARM state in case a limit switch is active."
You mean modify them so that they DO NOT return ALARM, don't you?.

**Comment by reszelaz** [comment]
Exactly! I will edit the comment. Thanks!

**Comment by reszelaz** [comment]
Agreeing on this solution will change the way how the clients should be written. If for example, a macro was doing:

~~~~
motion = self.getMotion(["mot01", mot02])
state, position = motion.move([10, 10])
if state == PyTango.DevState.ALARM:
    raise Exception("My motion ended in alarm, it could be a limit switch...")
~~~~

With this change, it will require checking the state of the limit switches explicitly. One of this type of implementation is the `mv` & co. macros. They should be adapted in order to show the information about the limit switches.

**Comment by reszelaz** [comment]
> the present documentation does not say anything about it.

This is not true. I have just found it documented in "The states" chapter of the Motor Tango class [documentation](http://www.sardana-controls.org/en/latest/devel/api/sardana/tango/pool/Motor.html#sardana.tango.pool.Motor.Motor).

**Comment by rhomspuron** [comment]
We changed the IcePAPCtrl to verify before if it is moving. The IcePAP protects the movement, for that reason is not necessary to check it on Sardana.
